litellm/caching.py

```diff
@@ -1446,12 +1449,389 @@ def delete_cache(self, key):
+            },
+            "fields": {
+                "text": [
+                    {"name": "response"},
+                    {"name": "prompt"},
+                    {"name": "response_hash"},
+                ],
+                "vector": [
+                    {
+                        "name": "litellm_embedding",
+                        "dims": self.embedding_vector_length,
+                        "distance_metric": "cosine",
+                        "algorithm": "flat",
+                        "datatype": "float32",
+                    }
+                ],
+            },
+        }
+
+        if redis_url is None:
+            # if no url passed, check if host, port and password are passed, if not raise an Exception
+            if host is None or port is None or password is None:
+                # try checking env for host, port and password
+                import os
+
+                host = os.getenv("REDIS_HOST")
+                port = os.getenv("REDIS_PORT")
+                password = os.getenv("REDIS_PASSWORD")
+                if host is None or port is None or password is None:
+                    raise Exception("Redis host, port, and password must be provided")
+
+            redis_url = "redis://:" + password + "@" + host + ":" + port
+        print_verbose(f"dual-semantic-cache redis_url: {redis_url} ({use_async})")
+        if not use_async:
+            self.index = SearchIndex.from_dict(schema)
+            self.index.connect(redis_url=redis_url)
+            try:
+                self.index.create(overwrite=False)
+
+            except Exception as e:
+                print_verbose(
+                    f"dual-semantic-cache: Got exception creating semantic cache index: {str(e)}"
+                )
+        else:
+            schema["index"]["name"] = (
+                f"litellm_semantic_cache_{embedding_vector_length}_index_async"
+            )
+            self.index = SearchIndex.from_dict(schema)
+            self.index.connect(redis_url=redis_url, use_async=True)
+
+    def _get_cache_logic(self, cached_response: Any):
+        """
+        Common 'get_cache_logic' across sync + async redis client implementations
+        """
+        if cached_response is None:
+            return cached_response
+
+        # check if cached_response is bytes
+        if isinstance(cached_response, bytes):
+            cached_response = cached_response.decode("utf-8")
+
+        try:
+            cached_response = json.loads(
+                cached_response
+            )  # Convert string to dictionary
+        except:
+            cached_response = ast.literal_eval(cached_response)
+        return cached_response
+
+    def set_cache(self, key, value, **kwargs):
+        import numpy as np
+
+        print_verbose(f"dual-semantic-cache (sync) set_cache, kwargs: {kwargs}")
+
+        messages = kwargs["messages"]
+        prompt = "".join(message["content"] for message in messages)
```

Comment:
the value of content in messages isn't always a string

i'd recommend using 'get_formatted_prompt' to handle this - https://github.com/BerriAI/litellm/blob/3167bee25aaae02f166c5048931d752580e10042/litellm/utils.py#L6626

Reply:
in the context of the semantic cache, i don't have access to the call type, which i think get_formatted_prompt needs. to handle cases where the content isn't a string, i'll add a bit of error handling (i'll also add the same error handling to the RedisSemanticCache, which this is based off of)